### DISCUSSION TITLE
Структура данных для API (JSON)

### DISCUSSION THREAD

**Discussion by SAKorney** [hhru-school/checkup-extension#17] [Ideas]
### Обсуждение состава и структуры запросов и ответов
**_Страница с описанием:_** https://github.com/hhru-school/checkup-extension/wiki/JSON

**Comment by SAKorney** [comment]
**Общее** 

1. Ознакомится с заметками Entities и Public (External) API в Miro
2. Добавить текстовую аннотацию назначения структур
3. Разделить запросы на две секции - поступление (соревнование) и администрирование 

**User**

Уточнение, проверка пользователей производится на стороне Checkup, вызовом соответствующей ручки на нашем back'e с токеном пользователя. Таким образом, наш frontend будет работать с токен пользователя как идентификатором. Вероятнее всего, он будет зашиваться в cookie.

Из информации, которую следует передавать по токену на frontend это:
1. Ф.И.О
4. Email

Определять доступные действия пользователя будем на backend и передавать роль на сторону клиента причины нет.

**Tasks**
"success": true - это предложение зашить HTTP-код ответа в данные ответа?
Требуется появнения за что отвечают поля "active", "step" и "stepStatus"?

**Task**
Требуется пояснение о назначении полей "content", "htmlTemplate", "cssTemplate", "jsTemplate"

**Solution history**
Это для отображения истории отправки решений на проверку?

**Solution**
Поля "htmlContent", "cssContent" и "jsContent" это решение от пользователя?
Предлагаю добавить поле Типа задачи, а эти три объединить и сделать вложенными в другое поле (см. Miro).

**Send solution** 
Post-запрос имеет вид "/api/resource". В нашем случае, это будет просто "/solution/".
Все данные в post-запросе обычно передаются в теле запроса. (Исключением будет токен пользователя, который будет в coockie).
В теле потребуется передать идентификатор решаемой задачи и само решение.  В случае с ответом будут следующие ситуации:

1. Первичная отправка (достаточно вернуть статус, что решение принято на проверку)
2. Проверка состояния решения (аналогично, достаточно будет статуса)
3. Повторный запрос решения полностью (например, когда пользователь перезагрузил страницу или хочет просмотреть прошлые попытки)

Из эти трёх, говоря о первичной отправки решения, относится только первый.

 **Result(s)**
Требуется пояснение по их назначениям.

**New task**
Необходимо решить со сценариями работы по созданию и редактированию задач (см. Miro заметка "Public (External) API").

**Reply by rmunirov** [reply]
> Вероятнее всего, он будет зашиваться в cookie.

В таком случае передаем токен 1 раз, при запросе информации о пользователе.

> Определять доступные действия пользователя будем на backend и передавать роль на сторону клиента причины нет.

Как, в таком случае, определять какую страницу показывать пользователю?

> "success": true - это предложение зашить HTTP-код ответа в данные ответа?

Да, чтобы можно было в `messages` закинуть более подробную информацию.

> Требуется появнения за что отвечают поля "active", "step" и "stepStatus"?

`active` - взято из макета, активность задачи - получаем возможность оперативно отключить определенную задачу из контеста, например, если обнаружили неточности, или решили заменить на другую
`step` - показывает статус последнего отправленного решения (в процессе решения - если решений еще нет, проверяется или получен результат - принято или отклонено) - нужно для отображения статуса по задаче
`stepStatus` - так же нужен для статуса, чтобы можно было подсветить характер самого статуса - все хорошо (синий) или все плохо (красный)

>   Требуется пояснение о назначении полей "content", "htmlTemplate", "cssTemplate", "jsTemplate"

`content` - это само описание задачи, что надо будет сделать в markdown разметке
`htmlTemplate` - это базовый шаблон решения, который будет вставляться в редактор, тут возможно он будет всегда одинаковый, стандартные теги, но это не точно
`cssTemplate` - так же базовый шаблон со стилями, возможно какие то уже будут прописаны, возможно будут приведены имена классов и тп
`jsTemplate` - базовый шаблон, возможно с сигнатурой функции и тп

> Это для отображения истории отправки решений на проверку?

Да, пользователю будет отображаться список уже отправленных решений, чтобы он мог их просмотреть, что то скопировать может быть и тп. Можно будет подумать над ограничением отправки, когда последнее решение было успешно принято.

> Поля "htmlContent", "cssContent" и "jsContent" это решение от пользователя?

Да, это то, что отправил пользователь на проверку

> Предлагаю добавить поле Типа задачи, а эти три объединить и сделать вложенными в другое поле (см. Miro).

По идее особой разницы пока что нет, можем и объединить.
По типу задачи - возможно избыточно, информация будет запрашиваться по определенной задаче, но наверное стоит добавить, возможно пригодится

> Все данные в post-запросе обычно передаются в теле запроса.

Согласен, чет загнался

> В случае с ответом будут следующие ситуации:

После отправки решения, его надо будет положить в историю, поэтому все таки лучше заполнить все поля и передать. Так же надо бы добавить запрос-пинг состояния решения, чтобы обновлять статус и показать его пользователю.

> Требуется пояснение по их назначениям.

Предполагались результаты по каждому пользователю, т.е. прошел или нет, возможно еще добавить какой статистики (с какого раза, % соответствия и прочее) на стороне checkUp это будет реализовано? Думаю было бы интереснее сделать это у нас...

> Необходимо решить со сценариями работы по созданию и редактированию задач (см. Miro заметка "Public (External) API").

Да, тут надо будет додумать.

**Reply by SAKorney** [reply]
> В таком случае передаем токен 1 раз, при запросе информации о пользователе.

Токен сессионый. Время его жизни может закончится в любой момент.

> Как, в таком случае, определять какую страницу показывать пользователю?

Уточни у Ромы. Исходя из надёжности, вид интерфейса, т.е. что показывать пользователю, лучше определять на стороне сервера и возвращать нужное представление, чтобы особо ушлые не лазили в разметку.

> stepStatus

Статус статуса? На первый взгляд, выглядит как дублирование, поскольку эту же информации можно получить из `step`. Если считаешь, что это оптимальнее сделать на бэке, сделаем.

> Template

Возможно, будет ли более оптимально сделать их опциональными в зависимости от типа задач? Все равно ведь нужно будет анализировать тип задачи. Или если это будет конкретная форма с заточенным обработчиком, тогда вообще не нужно передавать избыточные, пустые для такой формы поля. Что будут представлять из себя шаблоны, пока не понятно. Нужны примеры.

> Да, это то, что отправил пользователь на проверку

Снова предлагаю сделать вариативными в зависимости от типа задачи и передовать лишь необходимые

> По идее особой разницы пока что нет, можем и объединить.

Окей, вчера обсудили. Бэк будет выдавать линейные структуры без иерархий.

> После отправки решения, его надо будет положить в историю, поэтому все таки лучше заполнить все поля и передать. Так же надо бы добавить запрос-пинг состояния решения, чтобы обновлять статус и показать его пользователю.

Всё так. Здесь речь была о том, что для "пингующих" запросов не нужно гонять туда-обратно всю информацию.

 > Предполагались результаты по каждому пользователю, т.е. прошел или нет, возможно еще добавить какой статистики

Идея понятна. Оставим её на закуску? Успеем - сделаем. Всё-таки в приоритете функционал от заказчика. Но идея, да, хорошая.

**Reply by rmunirov** [reply]
> Возможно, будет ли более оптимально сделать их опциональными в зависимости от типа задач? Все равно ведь нужно будет анализировать тип задачи. Или если это будет конкретная форма с заточенным обработчиком, тогда вообще не нужно передавать избыточные, пустые для такой формы поля. Что будут представлять из себя шаблоны, пока не понятно. Нужны примеры.

Да, можно сделать их опциональными

**Comment by SAKorney** [comment]
Привёл образец как можно оформить структуры запросов/ответов на примере запроса получения задачи
